File: ruby/lib/runapi/runway/resources/extend_video.rb
# frozen_string_literal: true

module RunApi
  module Runway
    module Resources
      # Runway extend-video resource.
      # Append additional footage to a previously generated video, continuing from where the source task left off.
      class ExtendVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/runway/extend_video"
        RESPONSE_CLASS = Types::TaskCreateResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedTaskResponse
        MODEL = "runway"

        def initialize(http)
          @http = http
        end

        # Extend a video and wait until complete.
        #
        # @param source_task_id [String] ID of the completed TextToVideo or ExtendVideo task to continue from
        # @param prompt [String] prompt describing the continuation footage
        # @param output_resolution [String] must match the resolution of the source task ("720p" or "1080p")
        # @param watermark [String, nil] watermark text burned into the output
        # @param callback_url [String, nil] webhook URL for completion notification
        # @return [RunApi::Runway::Types::CompletedTaskResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create an extend-video task without waiting for completion.
        #
        # @param params [Hash] extend-video parameters (see {#run} for details)
        # @return [RunApi::Runway::Types::TaskCreateResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["extend-video"], params.merge(model: MODEL))
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get extend-video task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Runway::Types::TaskResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end

File: python/src/runapi/runway/resources/extend_video.py
"""Runway extend-video resource."""

from __future__ import annotations

from typing import Any

from runapi.core import Resource

from ..contract_gen import CONTRACT
from ..types import (
    CompletedTaskResponse,
    TaskCreateResponse,
)


class ExtendVideo(Resource):
    """Extend an existing Runway video task."""

    ENDPOINT = "/api/v1/runway/extend_video"

    RESPONSE_CLASS = TaskCreateResponse
    COMPLETED_RESPONSE_CLASS = CompletedTaskResponse

    MODEL = "runway"

    def run(self, **params: Any) -> Any:
        """Append footage to a previous video and poll until it completes.

        Args:
            **params: extend-video parameters (model, ...).

        Returns:
            The completed (narrowed) extend-video response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create an extend-video task and return immediately with an id.

        Args:
            **params: extend-video parameters (model, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["extend-video"], {**compacted, "model": self.MODEL})
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of an extend-video task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")
